fiat-core/src/main/java/com/netflix/spinnaker/fiat/providers/FileBasedApplicationProvider.java

```diff
@@ -0,0 +1,171 @@
+
+  private Thread watchThread;
+
+  public FileBasedApplicationProvider(){}
+
+  @VisibleForTesting
+  FileBasedApplicationProvider(Map<String, Application> applications) {
+    this.applications = applications;
+  }
+
+  @Override
+  public Set<Application> getApplications(@NonNull Collection<String> groups) {
+    return applications
+        .values()
+        .stream()
+        .filter(application ->
+                    application.getRequiredGroupMembership().isEmpty() ||
+                        !Collections.disjoint(application.getRequiredGroupMembership(), groups))
+        .collect(Collectors.toSet());
+  }
+
+  public FileBasedApplicationProvider watch(String watchedDirectory) {
+    watchThread = new Thread(new ConfigDirectoryWatcher(watchedDirectory));
+    watchThread.start();
+    return this;
+  }
+
+  @PreDestroy
+  public void close() {
+    if (watchThread != null && watchThread.isAlive()) {
+      watchThread.interrupt();
+    }
+  }
+
+  @RequiredArgsConstructor
+  private class ConfigDirectoryWatcher implements Runnable {
+
+    private final String watchedDirectory;
+
+    @Override
+    public void run() {
+      Path watchedDirPath = Paths.get(watchedDirectory);
+      parseExistingConfigFiles(watchedDirPath);
+      watchForChanges(watchedDirPath);
+    }
+
+    private void parseExistingConfigFiles(Path watchedDirPath) {
+      try (DirectoryStream<Path> dirStream = Files.newDirectoryStream(watchedDirPath)) {
+        for (Path config : dirStream) {
+          if (!isConfigFile(config.getFileName().toString())) {
+            continue;
+          }
+
+          parseApplicationConfig(config)
+              .ifPresent(app -> {
+                applications.put(app.getName(), app);
+                log.info("Adding configured application: " + app.getName());
+              });
+        }
+      } catch (IOException ioe) {
+        log.error("Error reading existing application configuration files", ioe);
+      }
+    }
+
+    private void watchForChanges(Path watchedDirPath) {
+      try (WatchService watchService = watchedDirPath.getFileSystem().newWatchService()) {
+        watchedDirPath.register(watchService,
+                                StandardWatchEventKinds.ENTRY_CREATE,
+                                StandardWatchEventKinds.ENTRY_MODIFY,
+                                StandardWatchEventKinds.ENTRY_DELETE);
+        log.info("Watching " + watchedDirectory + " for file changes");
+        while (true) {
+          try {
+            WatchKey watchKey = watchService.take();
+
+            for (WatchEvent event : watchKey.pollEvents()) {
+              String filepath = event.context().toString();
+              if (!isConfigFile(filepath)) {
+                continue;
+              }
+
+              switch (event.kind().toString()) {
+                case "ENTRY_CREATE":
+                case "ENTRY_MODIFY":
+                  val fullConfigPath = watchedDirPath.resolve((Path) event.context());
```

Comment:
```
case "ENTRY_CREATE":
case "ENTRY_MODIFY":
``` 

is a switch fall through - meaning execute whatever code is below up until the next `break` for either ENTRY_CREATE or ENTRY_MODIFY. Adding a log statement...